inttest/cplb-userspace/cplbuserspace_test.go

```diff
@@ -0,0 +1,223 @@
+
+		// Note that the token is intentionally empty for the first controller
+		s.Require().NoError(s.InitController(idx, "--config=/tmp/k0s.yaml", "--disable-components=metrics-server", "--enable-worker", joinToken))
+		s.Require().NoError(s.WaitJoinAPI(s.ControllerNode(idx)))
+
+		// With the primary controller running, create the join token for subsequent controllers.
+		if idx == 0 {
+			token, err := s.GetJoinToken("controller")
+			s.Require().NoError(err)
+			joinToken = token
+		}
+	}
+
+	// Final sanity -- ensure all nodes see each other according to etcd
+	for idx := range s.BootlooseSuite.ControllerCount {
+		s.Require().Len(s.GetMembers(idx), s.BootlooseSuite.ControllerCount)
+	}
+
+	// Create a worker join token
+	workerJoinToken, err := s.GetJoinToken("worker")
+	s.Require().NoError(err)
+
+	// Start the workers using the join token
+	s.Require().NoError(s.RunWorkersWithToken(workerJoinToken))
+
+	client, err := s.KubeClient(s.ControllerNode(0))
+	s.Require().NoError(err)
+
+	for idx := range s.BootlooseSuite.ControllerCount {
+		s.Require().NoError(s.WaitForNodeReady(s.ControllerNode(idx), client))
+	}
+	s.Require().NoError(s.WaitForNodeReady(s.WorkerNode(0), client))
+
+	// Verify that none of the servers has the dummy interface
+	for idx := range s.BootlooseSuite.ControllerCount {
+		s.checkDummy(ctx, s.ControllerNode(idx))
+	}
+
+	// Verify that only one controller has the VIP in eth0
+	count := 0
+	for idx := range s.BootlooseSuite.ControllerCount {
+		if s.hasVIP(ctx, s.ControllerNode(idx), lb) {
+			count++
+		}
+	}
+	s.Require().Equal(1, count, "Expected exactly one controller to have the VIP")
+
+	// Verify that controller+worker nodes are working normally.
+	s.T().Log("waiting to see CNI pods ready")
+	s.Require().NoError(common.WaitForKubeRouterReady(s.Context(), client), "kube router did not start")
+	s.T().Log("waiting to see konnectivity-agent pods ready")
+	s.Require().NoError(common.WaitForDaemonSet(s.Context(), client, "konnectivity-agent", "kube-system"), "konnectivity-agent did not start")
+	s.T().Log("waiting to get logs from pods")
+	s.Require().NoError(common.WaitForPodLogs(s.Context(), client, "kube-system"))
+
+	s.T().Log("Testing that the load balancer is actually balancing the load")
+	// Other stuff may be querying the controller, running the HTTPS request 15 times
+	// should be more than we need.
+	signatures := make(map[string]int)
+	url := url.URL{Scheme: "https", Host: net.JoinHostPort(lb, strconv.Itoa(6443))}
+	for range 15 {
+		signature, err := getServerCertSignature(url.String())
+		s.Require().NoError(err)
+		signatures[signature] = 1
+	}
+
+	s.Require().Len(signatures, 3, "Expected 3 different signatures, got %d", len(signatures))
+}
+
+// getLBAddress returns the IP address of the controller 0 and it adds 100 to
+// the last octet unless it's bigger or equal to 154, in which case it
+// subtracts 100. Theoretically this could result in an invalid IP address.
```

Comment:
Modified it for:

```
// getLBAddress returns the IP address of the controller 0 and it adds 100 to
// the last octet unless it's bigger or equal to 154, in which case it
// subtracts 100. Theoretically this could result in an invalid IP address.
// This is so that get a virtual IP in the same subnet as the controller.

```